src/libraries/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs

```diff
@@ -149,79 +151,82 @@ internal void InitAssemblyMethods(XmlMapping[] xmlMappings)
             contract = null;
             string? serializerName = null;
 
-            // check to see if we loading explicit pre-generated assembly
-            object[] attrs = type.GetCustomAttributes(typeof(System.Xml.Serialization.XmlSerializerAssemblyAttribute), false);
-            if (attrs.Length == 0)
+            using (AssemblyLoadContext.EnterContextualReflection(type.Assembly))
             {
-                // Guess serializer name: if parent assembly signed use strong name
-                AssemblyName name = type.Assembly.GetName();
-                serializerName = Compiler.GetTempAssemblyName(name, defaultNamespace);
-                // use strong name
-                name.Name = serializerName;
-                name.CodeBase = null;
-                name.CultureInfo = CultureInfo.InvariantCulture;
-
-                try
-                {
-                    serializer = Assembly.Load(name);
-                }
-                catch (Exception e)
-                {
-                    if (e is OutOfMemoryException)
+                // check to see if we loading explicit pre-generated assembly
+                object[] attrs = type.GetCustomAttributes(typeof(System.Xml.Serialization.XmlSerializerAssemblyAttribute), false);
+                if (attrs.Length == 0)
+                {
+                    // Guess serializer name: if parent assembly signed use strong name
+                    AssemblyName name = type.Assembly.GetName();
+                    serializerName = Compiler.GetTempAssemblyName(name, defaultNamespace);
+                    // use strong name
+                    name.Name = serializerName;
+                    name.CodeBase = null;
+                    name.CultureInfo = CultureInfo.InvariantCulture;
+
+                    try
                     {
-                        throw;
+                        serializer = Assembly.Load(name);
```

Comment:
Instead of using contextual reflection it would be cleaner to just call `AssemblyLoadContext.GetLoadContext(type.Assembly).LoadFromAssemblyName`.

We recommend people stop using the `Assembly.Load` because of its weird behavior in terms of which ALC it loads into (without contextual reflection it relies on caller's ALC which is often times wrong).